### PR TITLE
Implemented insertion sort

### PR DESCRIPTION
**Pull request by joshua041904**
I implemented insertion.ts, which contains the implementation for insertion sort and it's corresponding visualization.

Made a few edits to ArrayInput.vue (importing insertion.ts and added insertion sort pseudo code). 

Made minimal edits to bubble.ts (exported swap and delayExecution functions so I could use them in insertion.ts)

### CODE REVIEW

**Comment by Yonava** [comment]
I'm seeing some good improvements. The title of the PR is good, the size and scope is perfect, for the description, you dont need to tell me which files you changed because I can see that, tell me what the feature you added was or bug you patched etc.

In general I see 3 areas of opportunity:
* Practice your JS mental models in order to reduce the amount of unnecessary function arguments and clean up the code
* Work on styling your code consistently, I recommend installing `prettier` and configuring vscode to format on save. You can find tutorials online, but if you are having trouble, I can help get you set up. This will save you tons of time and make your code look 10x better
* reduce the amount of comments you are writing, and focus on the clarity of your function and variable names instead. A rule of thumb, imagine you are simply interested in using a function without being concerned with reading the internals, would you understand how your inputs correspond to the functions output?

⭐ ⭐ ⭐

File: codebase/src/sorting/algorithms/bubble.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ const swap = (
     });
 };
 
-const delayExecution = (delay: number) => {
+export const delayExecution = (delay: number) => {
     return new Promise<void>((resolve) => {
         setTimeout(resolve, delay);
     });
```
[the remark this answers is not in the document]
Comment (joshua041904) [reply codebase/src/sorting/algorithms/bubble.ts]: Just fixed this and made a commit.

File: codebase/src/sorting/components/ArrayInput.vue [hunk 2]
```diff
@@ -127,6 +141,8 @@ const clearArray = () => {
 // Function to sort the array using bubble sort algorithm
 const sortArray = async () => {
     await bubbleSort(array.value, 500, updateSwap, updateLeftIndex, updateRightIndex, updateCurrentLines, updateSortedIndex);
```
[the remark this answers is not in the document]
Comment (joshua041904) [reply codebase/src/sorting/components/ArrayInput.vue]: Thanks. Just changed the bubbleSort and insertionSort functions to use the options object pattern.

File: codebase/src/sorting/algorithms/insertion.ts [hunk 3]
```diff
@@ -1,29 +1,64 @@
...
+import { swap, delayExecution } from "./bubble";
 
 // Function to sort array using insertion sort
-function insertionSort(arr: number[]) {
-    for (let i = 1; i < arr.length; i++) {
-        let key = arr[i];
-        let j = i - 1;
+export const insertionSort = async (
+    array: number[],
+    delay: number,
```
[the remark this answers is not in the document]
Comment (joshua041904) [reply codebase/src/sorting/algorithms/insertion.ts]: I think this was fixed when I switched insertionSort to options object pattern. delay is now ms

File: codebase/src/sorting/algorithms/insertion.ts [hunk 4]
```diff
@@ -1,29 +1,64 @@
...
+    updateSwap: (newArray: number[]) => void,
+    updateLeftIndex: (index: number | null) => void,
+    updateRightIndex: (index: number | null) => void,
+    updateCurrentLines: (lines: number[]) => void,
+    updateSortedIndex: (index: number | null) => void
+): Promise<number[]> => {
```
[the remark this answers is not in the document]
Comment (joshua041904) [reply codebase/src/sorting/algorithms/insertion.ts]: Is the return type implied for all these lines? Meaning I should delete all the "=> void"s and "Promise<number[]>"?

File: codebase/src/sorting/algorithms/insertion.ts [hunk 5]
```diff
@@ -1,29 +1,64 @@
...
-        /* Move elements of arr[0..i-1], that are
-           greater than key, to one position ahead
-           of their current position */
-        while (j >= 0 && arr[j] > key) {
-            arr[j + 1] = arr[j];
-            j = j - 1;
+            let j = i - 1;
+            // Highlight "while left_element >=  0"
+            updateCurrentLines([3]);
+            while(j >= 0){
```
[the remark this answers is not in the document]
Comment (joshua041904) [reply codebase/src/sorting/algorithms/insertion.ts]: Just set up the prettier extension to work on every save...so this should be good now.

File: codebase/src/sorting/components/ArrayInput.vue [hunk 6]
```diff
@@ -53,6 +53,7 @@
 <script setup lang="ts">
 import { ref, defineEmits } from 'vue';
 import { bubbleSort } from '../algorithms/bubble';
+import { insertionSort } from '../algorithms/insertion';
 
 // Define custom event to emit sorted array
 const emit = defineEmits(['sortedArray']);
```
[the remark this answers is not in the document]
Comment (joshua041904) [reply codebase/src/sorting/components/ArrayInput.vue]: Thanks, I just switched it to typed emits

**Pull request merged**